### PR TITLE
Make the library Multiplatform

### PR DESCRIPTION
**Pull request by AYastrebov**
## Make the library Multiplatform

This PR migrates kotlin-obd-api from a JVM-only library to [Kotlin Multiplatform](https://kotlinlang.org/docs/multiplatform.html), enabling use on iOS, macOS, Linux, Windows, JS, and Wasm — while preserving all existing behavior and tests.

### Build & Tooling
- Migrated from `kotlin("jvm")` to `kotlin("multiplatform")` plugin
- Added Gradle version catalog (`libs.versions.toml`) for centralized dependency management
- Added Develocity build scans (published on CI only)
- Updated Kotlin to 2.3.21, Gradle to 9.5.0, ktlint to 14.2.0, detekt to 2.0.0-alpha.3

### Platform Targets
- `jvm`, `iosArm64`, `iosSimulatorArm64`, `macosArm64`, `linuxX64`, `mingwX64`, `js(nodejs)`, `wasmJs(nodejs)`
- Removed deprecated targets: `iosX64`, `macosX64`

### Source Layout
- Moved `src/main/` → `src/commonMain/`, `src/test/` → `src/commonTest/`

### JVM API → Multiplatform Replacements
- `java.util.regex.Pattern` → Kotlin `Regex`
- `java.io.InputStream`/`OutputStream` → `kotlinx.io.Source`/`Sink`
- `String.format("%.Nf", ...)` → custom `formatFloat()` (integer modulo arithmetic)
- `"%02X".format(...)` → custom `formatHex()`
- `Integer.parseInt`/`toHexString` → Kotlin stdlib equivalents
- `javaClass.simpleName` → `this::class.simpleName`
- `measureTimeMillis` → `TimeSource.Monotonic`
- `Dispatchers.IO` → `Dispatchers.Default` (kotlinx-io is non-blocking)

### Test Migration
- JUnit `@RunWith(Parameterized::class)` → `kotlin.test` with inline test data
- Tests use production `formatFloat`/`formatHex` directly
- All 66 tests pass on JVM

### CI Improvements
- Develocity build scans for CI runs
- Gradle cache-read-only on PRs (prevents cache pollution)
- Concurrency group to cancel stale in-progress runs
- Test report artifacts uploaded (7-day retention)
- Full `./gradlew build` runs all targets

### Bug Fixes (pre-existing, found during review)
- **RuntimeCommand wrong PID**: `0F` → `1F` (Engine Runtime) — see #42
- **UNSUPPORTED_COMMAND_MESSAGE_PATTERN**: removed spaces so it matches — see #41
- **Cache key collision**: use `command.tag` to prevent collisions
- **Retry loop**: restored original `while(isReading)` structure
- **formatFloat precision**: integer modulo arithmetic avoids overflow

### Dependencies
- Added: `kotlinx-io-core` 0.9.0, `kotlinx-coroutines-test`
- Removed: `kotlin-stdlib-jdk8`, `kotlin-test-junit`

### CODE REVIEW

**Comment by ewilliamsfleetio** [comment]
@AYastrebov do you have any code examples? I'm trying to use this branch for some multiplatform shenanigans but I couldn't find a good example of converting Swift input/output into Kotlinx `Source` objects

**Comment by AYastrebov** [comment]
@ewilliamsfleetio not yet. I was planning to use this lib in my pet project, but unfortunately, I can't find enough time to do that.

**Comment by eltonvs** [comment]
Hey @AYastrebov, thank you for working on that.
I finally have some more time to work on this project, and making it KMP-compatible is my next focus.

Could you walk me through the possible breaking changes for this (for existing projects using the library)?
I'm currently targeting this change to v2.0.0.

Also, I recently fixed the coroutine usage in the lib + added a linter and fixed the issues, causing some conflicts with this branch.

**Comment by AYastrebov** [comment]
Hi, @eltonvs !
I will update this MR and write a short report

**Comment by AYastrebov** [comment]
# Migration to Kotlin Multiplatform

## Summary

The `multiplatform` branch converts kotlin-obd-api from a JVM-only library to Kotlin Multiplatform, targeting JVM, iOS, macOS, Linux, Windows, JS, and WebAssembly — in 9 linear commits.

## Breaking Changes for Existing Users

### 1. Connection constructor signature changed

```kotlin
// Before
ObdDeviceConnection(inputStream: InputStream, outputStream: OutputStream)

// After
ObdDeviceConnection(inputStream: Source, outputStream: Sink)
```

Users must convert their `InputStream`/`OutputStream` to kotlinx-io `Source`/`Sink`. On JVM, this is typically `inputStream.asSource()` and `outputStream.asSink()` from `kotlinx-io-jvm`.

### 2. New dependency: kotlinx-io

Projects must add `kotlinx-io-core` to their dependency graph. It's a transitive dependency of the library, but transport code that creates `Source`/`Sink` will need the explicit import.

### 3. `RegexPatterns` types changed

`Pattern` fields are now `Regex`. Anyone extending the library and referencing `RegexPatterns` directly (unlikely for most users) will need to update from `java.util.regex.Pattern` API calls to `kotlin.text.Regex`.

### 4. Exception `toString()` output slightly different

`javaClass.simpleName` → `::class.simpleName` — same value on JVM, but technically a different code path.

### 5. Float formatting may differ at edge cases

`"%.2f".format()` → custom `formatFloat()`. Output is identical for normal values but could differ for extreme floats (NaN, Infinity, very large numbers).

## What's NOT Breaking

- All command classes (`RPMCommand`, `SpeedCommand`, `VINCommand`, etc.) — unchanged API
- `ObdResponse` / `ObdRawResponse` — same fields and types
- `run()` method signature — same parameters and return type
- Cache behavior, mutex serialization, retry logic — identical

**Bottom line:** The only change most users will feel is wrapping their streams in `asSource()`/`asSink()` and adding the kotlinx-io dependency.

File: src/commonMain/kotlin/com/github/eltonvs/obd/connection/ObdDeviceConnection.kt [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            outputStream.flush()
+        }
+        if (delayTime > 0) {
+            delay(delayTime)
+        }
+    }
+
+    private suspend fun readRawData(maxRetries: Int): String =
+        withContext(dispatcher) {
+            val res = StringBuilder()
```
[the remark this answers is not in the document]
Comment (AYastrebov) [reply src/commonMain/kotlin/com/github/eltonvs/obd/connection/ObdDeviceConnection.kt:79]: This is a valid concern, but in practice `Source.request(1)` returning `false` does not exclusively mean permanent EOF — it can also mean the buffer is temporarily empty (depending on the `Source` implementation wrapping the transport). This mirrors the original `InputStream.available() == 0` behavior where no bytes are ready yet but more may arrive.

The existing test `retries when buffer is initially empty then data arrives` confirms this: it simulates a `Source` that returns `false` on the first `request(1)` call, then provides data on the next. Breaking immediately on `false` would fail this test.

Added a clarifying comment in ae759f6.

File: src/commonTest/kotlin/com/github/eltonvs/obd/FormatFloat.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        if (isNegative) append('-')
+        append(intPart)
+        append('.')
+        append(fracPart.toString().padStart(decimalPlaces, '0').take(decimalPlaces))
+    }
+}
+
+/** Format hex value with zero padding. */
+internal fun formatHex(value: Int): String {
+    val hex = value.toString(16).uppercase()
```
[the remark this answers is not in the document]
Comment (AYastrebov) [reply src/commonTest/kotlin/com/github/eltonvs/obd/FormatFloat.kt]: Good catch. Deleted the test-local `FormatFloat.kt` and updated all 8 test files to import directly from `com.github.eltonvs.obd.command.formatFloat`/`formatHex`. Tests now exercise the production code path. Fixed in 8dfa478.

File: src/commonMain/kotlin/com/github/eltonvs/obd/command/RegexPatterns.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    val STARTS_WITH_ALPHANUM_PATTERN: Regex = "[^a-z0-9 ]".toRegex(RegexOption.IGNORE_CASE)
+
+    // Error patterns
+    const val BUSINIT_ERROR_MESSAGE_PATTERN = "BUS INIT... ERROR"
+    const val MISUNDERSTOOD_COMMAND_MESSAGE_PATTERN = "?"
+    const val NO_DATE_MESSAGE_PATTERN = "NO DATA"
+    const val STOPPED_MESSAGE_PATERN = "STOPPED"
+    const val UNABLE_TO_CONNECT_MESSAGE_PATTERN = "UNABLE TO CONNECT"
+    const val ERROR_MESSAGE_PATTERN = "ERROR"
+    const val UNSUPPORTED_COMMAND_MESSAGE_PATTERN = "7F 0[0-A] 1[1-2]"
```
[the remark this answers is not in the document]
Comment (AYastrebov) [reply src/commonMain/kotlin/com/github/eltonvs/obd/command/RegexPatterns.kt]: On second thought, this is serious enough to fix here. Done in cf62222 (pattern fix) and d18313e (test coverage with both positive and negative cases). Filed as upstream issue #41.

File: src/commonMain/kotlin/com/github/eltonvs/obd/connection/ObdDeviceConnection.kt [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            var retriesCount = 0
+
+            // Read until '>' arrives or retries are exhausted.
+            // request(1) returns false when no bytes are buffered yet;
```
[the remark this answers is not in the document]
Comment (AYastrebov) [reply src/commonMain/kotlin/com/github/eltonvs/obd/connection/ObdDeviceConnection.kt]: Good catch. Restored the original `while (isReading)` loop structure — always attempts to read available data first, only consults `retriesCount` in the no-data branch. This matches the original `InputStream.available()` semantics exactly. Fixed in 8076d50.

**Comment by AYastrebov** [comment]
Hi @eltonvs ,
I've delivered some more updates, some bugs were discovered and fixed. 
I think it's ready to be merged now
